contracts/ethregistrar/ExponentialPremiumPriceOracle.sol

```diff
@@ -0,0 +1,143 @@
+pragma solidity >=0.8.4;
+
+import "./SafeMath.sol";
+import "./StablePriceOracle.sol";
+
+contract ExponentialPremiumPriceOracle is StablePriceOracle {
+    uint256 constant GRACE_PERIOD = 90 days;
+    uint256 constant START_PREMIUM = 100000000 * 1e18; // $100 mil start price
+    uint256 immutable endValue; // $0.372...
+
+    constructor(
+        AggregatorInterface _usdOracle,
+        uint256[] memory _rentPrices,
+        uint256 lastDay
+    ) StablePriceOracle(_usdOracle, _rentPrices) {
+        endValue = START_PREMIUM >> lastDay;
+    }
+
+    uint256 constant PRECISION = 1e18;
+    uint256 constant SECONDS_IN_DAY = 86400;
+    uint256 constant bit1 = 999989423469314432; // 0.5 ^ 1/65536 * (10 ** 18)
+    uint256 constant bit2 = 999978847050491904; // 0.5 ^ 2/65536 * (10 ** 18)
+    uint256 constant bit3 = 999957694548431104;
+    uint256 constant bit4 = 999915390886613504;
+    uint256 constant bit5 = 999830788931929088;
+    uint256 constant bit6 = 999661606496243712;
+    uint256 constant bit7 = 999323327502650752;
+    uint256 constant bit8 = 998647112890970240;
+    uint256 constant bit9 = 997296056085470080;
+    uint256 constant bit10 = 994599423483633152;
+    uint256 constant bit11 = 989228013193975424;
+    uint256 constant bit12 = 978572062087700096;
+    uint256 constant bit13 = 957603280698573696;
+    uint256 constant bit14 = 917004043204671232;
+    uint256 constant bit15 = 840896415253714560;
+    uint256 constant bit16 = 707106781186547584;
+
+    /**
+     * @dev Store value in variable
+     * @param startPremium value to store
+     * @param elapsed time past since expiry
+     */
+    function decayedPremium(uint256 startPremium, uint256 elapsed)
+        public
+        view
+        returns (uint256)
+    {
+        uint256 daysPast = (elapsed * PRECISION) / SECONDS_IN_DAY;
+        uint256 intDays = daysPast / PRECISION;
+        uint256 premium = startPremium >> intDays;
+        uint256 partDay = (daysPast - intDays * PRECISION);
+        uint256 fraction = (partDay * (2**16)) / PRECISION;
+        uint256 totalPremium = addFractionalPremium(fraction, premium);
+        return totalPremium;
+    }
+
+    /**
+     * @dev Returns the pricing premium in internal base units.
+     */
+    function _premium(
+        string memory,
+        uint256 expires,
+        uint256
+    ) internal view override returns (uint256) {
+        expires = expires + GRACE_PERIOD;
+        if (expires > block.timestamp) {
+            return 0;
+        }
+
+        uint256 elapsed = block.timestamp - expires;
+        uint256 premium = decayedPremium(START_PREMIUM, elapsed);
+        if (premium >= endValue) {
```

Comment:
It shouldn't be possible for this to be false - can we just remove the if?

Reply:
We don't set an end time for the premium, so the premium will keep decreasing after the last day, e.g. after day 21 it will be below $47 and so at that point, the premium would be say $40 -$47 which would be -$7 or just 0 if it saturates?

I get this when I remove it:

```
Error: VM Exception while processing transaction: reverted with panic code 0x11 (Arithmetic operation underflowed or overflowed outside of an unchecked block)
```

I am checking a day past the last day in my tests to make sure we get back 0. I guess the alternative would be to just revert, but that would be almost the same as an if statement,